src/stream/src/executor/arrangement_backfill.rs

```diff
@@ -0,0 +1,679 @@
+        // We construct a backfill stream with upstream as its left input and mv snapshot read
+        // stream as its right input. When a chunk comes from upstream, we will buffer it.
+        //
+        // When a barrier comes from upstream:
+        //  - Update the `snapshot_read_epoch`.
+        //  - For each row of the upstream chunk buffer, forward it to downstream if its pk <=
+        //    `current_pos`, otherwise ignore it.
+        //  - reconstruct the whole backfill stream with upstream and new mv snapshot read stream
+        //    with the `snapshot_read_epoch`.
+        //
+        // When a chunk comes from snapshot, we forward it to the downstream and raise
+        // `current_pos`.
+        //
+        // When we reach the end of the snapshot read stream, it means backfill has been
+        // finished.
+        //
+        // Once the backfill loop ends, we forward the upstream directly to the downstream.
+        if to_backfill {
+            let mut upstream_chunk_buffer: Vec<StreamChunk> = vec![];
+            let mut pending_barrier: Option<Barrier> = None;
+            'backfill_loop: loop {
+                let mut cur_barrier_snapshot_processed_rows: u64 = 0;
+                let mut cur_barrier_upstream_processed_rows: u64 = 0;
+
+                // Process barrier
+                // Each time we break out of the backfill_stream loop,
+                // it is typically due to barrier (except the first time).
+                // So we should immediately process barrier if there's one pending.
+                if let Some(barrier) = pending_barrier.clone() {
+                    // If it is a barrier, switch snapshot and consume
+                    // upstream buffer chunk
+
+                    if let Some(current_pos) = &current_pos {
+                        // NOTE: We do not consume the chunk here.
+                        // We still need to flush it to the `state_table`,
+                        // to ensure the upstream state is replicated.
+                        for chunk in &upstream_chunk_buffer {
+                            cur_barrier_upstream_processed_rows += chunk.cardinality() as u64;
+                            yield Message::Chunk(Self::mapping_chunk(
+                                Self::mark_chunk(
+                                    chunk,
+                                    current_pos,
+                                    &pk_in_output_indices,
+                                    &pk_order,
+                                ),
+                                &self.output_indices,
+                            ));
+                        }
+                    }
+                    if !upstream_chunk_buffer.is_empty() {
+                        for chunk in upstream_chunk_buffer.drain(..) {
+                            upstream_table.write_chunk(chunk);
+                        }
+                        upstream_table.commit(barrier.epoch).await?;
+                    }
+
```

Comment:
It seems we could iterate these two for loop at the same time, that is yielding the chunk and write it to the `upstream_table`.